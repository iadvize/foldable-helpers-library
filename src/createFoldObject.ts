/* eslint-disable @typescript-eslint/no-explicit-any */
import * as ArrayFP from 'fp-ts/es6/Array';
import * as Either from 'fp-ts/es6/Either';
import * as RecordFP from 'fp-ts/es6/Record';
import { pipe } from 'fp-ts/es6/pipeable';
import { identity } from 'fp-ts/es6/function';

import { Guard, Func } from './types';

const DEFAULT_CLAUSE = '_';

type Guards<TypesRecord extends Record<string, any>> = {
  [key in keyof TypesRecord]: Guard<any, TypesRecord[key]>;
};

export function createFoldObject<TypesRecord extends Record<string, any>>(
  guards: Guards<TypesRecord>,
) {
  type FullFoldFunctions<R> = {
    [key in keyof TypesRecord]: Func<TypesRecord[key], R>;
  };

  type PartialFoldFunctions<R> = Partial<FullFoldFunctions<R>>;

  type OmittedTypesRecord<R> = Omit<TypesRecord, keyof PartialFoldFunctions<R>>;

  type OmittedTypesInstance<R> = OmittedTypesRecord<R>[keyof OmittedTypesRecord<
    R
  >];

  type PartialFoldFunctionsWithDefault<R> = PartialFoldFunctions<R> & {
    _: Func<OmittedTypesInstance<R>, R>;
  };

  type FoldFunctions<R> =
    | PartialFoldFunctionsWithDefault<R>
    | FullFoldFunctions<R>;

  type Instance = TypesRecord[keyof TypesRecord & string];

  return function fold<R>(funcs: FoldFunctions<R>) {
    return (s: Instance): R => {
      const keys = RecordFP.keys(guards);

      return pipe(
        keys,
        ArrayFP.findFirst(key => guards[key](s)),
        Either.fromOption(() => new Error(`No guard found to fold ${s}`)),
        Either.chain(key => {
          if (key in funcs) {
            const func = funcs[key];

            if (typeof func !== 'function') {
              return Either.left(
                new Error(`"${key}" clause should be a function`),
              );
            }

            return Either.right(func(s));
          }

          if (DEFAULT_CLAUSE in funcs) {
            const func = funcs[DEFAULT_CLAUSE];

            if (typeof func !== 'function') {
              return Either.left(
                new Error(`"${DEFAULT_CLAUSE}" clause should be a function`),
              );
            }

            // s is not refined magically by TS so we have to force the type
            const localS = s as OmittedTypesInstance<R>;

            return Either.right(func(localS));
          }

          return Either.left(
            new Error(`No "${key}" nor "_" clause passed to fold`),
          );
        }),
        Either.fold(
          // if no function found, we throw error, guards or fold clauses are
          // broken
          error => {
            throw error;
          },
          // we found a function, we return the result of the function applied
          // to the given instance
          identity,
        ),
      );
    };
  };
}
