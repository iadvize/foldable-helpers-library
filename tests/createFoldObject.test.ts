import { pipe } from 'fp-ts/es6/pipeable';
import { identity } from 'fp-ts/es6/function';
import { map, range } from 'fp-ts/es6/Array';

import { createFoldObject } from '../src';

type Type<Tag> = {
  tag: Tag;
};

function createIsType<Tag>(tag: Tag) {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return function isType(t: { tag: any }): t is Type<Tag> {
    return t.tag === tag;
  };
}

type Type1 = Type<1>;
type Type2 = Type<2>;

const isType1 = createIsType(1 as const);
const isType2 = createIsType(2 as const);
const isType3 = createIsType(3 as const);
const isType4 = createIsType(4 as const);
const isType5 = createIsType(5 as const);
const isType6 = createIsType(6 as const);

describe('createFoldObject', () => {
  it('created fold should call the corresponding function for 2 types', () => {
    const fold = createFoldObject({
      type1: isType1,
      type2: isType2,
    });

    const type1: Type1 = { tag: 1 };
    const result1 = pipe(
      type1,
      fold({
        type1: () => 'result1',
        type2: () => 'result2',
      }),
    );

    expect(result1).toEqual('result1');

    const type2: Type2 = { tag: 2 };
    const result2 = pipe(
      type2,
      fold({
        type1: () => 'result1',
        type2: () => 'result2',
      }),
    );

    expect(result2).toEqual('result2');
  });

  it('created fold should call the corresponding function for 6 types', () => {
    const fold = createFoldObject({
      type1: isType1,
      type2: isType2,
      type3: isType3,
      type4: isType4,
      type5: isType5,
      type6: isType6,
    });

    const foldFunction = fold({
      type1: () => 'result1',
      type2: () => 'result2',
      type3: () => 'result3',
      type4: () => 'result4',
      type5: () => 'result5',
      type6: () => 'result6',
    });

    const tags = range(1, 6) as (1 | 2 | 3 | 4 | 5 | 6)[];

    pipe(
      tags,
      map(t => {
        const oneType = { tag: t };

        const result = pipe(oneType, foldFunction);

        expect(result).toEqual(`result${t}`);
      }),
    );
  });

  it('created fold should throw if no guard found', () => {
    const fold = createFoldObject({
      type1: isType1,
      type2: isType2,
    });

    const tag = { tag: 3 };
    const someFunc = fold<{ tag: number }>({
      type1: identity,
      type2: identity,
    });

    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    const funcThatWillThrow = () => someFunc(tag);

    expect(funcThatWillThrow).toThrowErrorMatchingSnapshot();
  });

  it('created fold should accept a default clause', () => {
    const fold = createFoldObject({
      type1: isType1,
      type2: isType2,
      type3: isType3,
      type4: isType4,
      type5: isType5,
      type6: isType6,
    });

    const result1 = pipe(
      { tag: 1 },
      fold({
        type1: () => 'result1',
        _: () => 'default',
      }),
    );

    expect(result1).toEqual(`result1`);

    const result2 = pipe(
      { tag: 2 },
      fold({
        type1: () => 'result1',
        _: () => 'default',
      }),
    );

    expect(result2).toEqual(`default`);
  });

  it('created fold should throw if clause function found', () => {
    const fold = createFoldObject({
      type1: isType1,
      type2: isType2,
    });

    const tag = { tag: 2 };
    const someFunc = fold<{ tag: number }>({
      type1: identity,
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore - testing behavior without type safety
      type3: identity,
    });

    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    const funcThatWillThrow = () => someFunc(tag);

    expect(funcThatWillThrow).toThrowErrorMatchingSnapshot();
  });

  it('created fold should throw if clause is not a function', () => {
    const fold = createFoldObject({
      type1: isType1,
      type2: isType2,
    });

    const tag = { tag: 2 };
    const someFunc = fold<{ tag: number }>({
      type1: identity,
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore - testing behavior without type safety
      type2: 'toto',
    });

    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    const funcThatWillThrow = () => someFunc(tag);

    expect(funcThatWillThrow).toThrowErrorMatchingSnapshot();
  });

  it('created fold should throw if default clause is not a function', () => {
    const fold = createFoldObject({
      type1: isType1,
      type2: isType2,
    });

    const tag = { tag: 2 };
    const someFunc = fold<{ tag: number }>({
      type1: identity,
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore - testing behavior without type safety
      _: 'toto',
    });

    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    const funcThatWillThrow = () => someFunc(tag);

    expect(funcThatWillThrow).toThrowErrorMatchingSnapshot();
  });
});
