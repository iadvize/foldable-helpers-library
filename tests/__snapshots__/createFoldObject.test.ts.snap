// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createFoldObject created fold should throw if clause function found 1`] = `"No \\"type2\\" nor \\"_\\" clause passed to fold"`;

exports[`createFoldObject created fold should throw if clause is not a function 1`] = `"\\"type2\\" clause should be a function"`;

exports[`createFoldObject created fold should throw if default clause is not a function 1`] = `"\\"_\\" clause should be a function"`;

exports[`createFoldObject created fold should throw if no guard found 1`] = `"No guard found to fold [object Object]"`;
